messagecatalog/spec.md

```diff
@@ -0,0 +1,1475 @@
+or infrastructure. An event is a kind of message.
+
+A producer produces messages and publishes them. A subscriber subscribes to
+published messages based on some criteria. A consumer is the delivery
+destination for such subscriptions and handles the delivered information.
+
+## 1. Introduction
+
+Message and event catalogs are useful tools at design and development time
+because they allow interactive discovery of the kinds of messages that are
+available for subscription from a system and to find out, for instance, which
+event sources raise them. A chosen message definition catalog entry can serve as
+the basis for generating strongly typed representations of the message or other
+handling code.
+
+The message catalog defined here contains groups of abstract message
+definitions. A group is typically associated with a kind of producer. The
+message definitions belonging to the group describe messages that the
+producer(s) might send. The catalog can store multiple versions of a message
+definition.
+
+When a defined message is meant to carry a payload, this specification allows
+for references to an external schema document or to embed such a document. The
+external reference MAY refer to a CNCF Schema Registry endpoint.
+
+Structurally and functionally, this message catalog API is very similar to the
+CNCF Schema Registry API, and it is reasonable and desirable for those APIs to
+be collocated in the same service and on the same endpoint.
+
+The Message Catalog and Schema Registry data formats and APIs exist in parallel
+because they serve different purposes. The Schema Registry is a general purpose
+store for serialization and validation schemas documents for structured data
+that is useful for describing message and event payloads, but also in all other
+scenarios where structured data is handled. For instance, a data structure might
+be sent to a system inside of a message but might subsequently be stored inside
+a data lake without the message envelope. The Message Catalog is more
+constrained and focused on message definitions, which covers both payload and
+metadata of the message. In environments where data structures are shared across
+multiple services and applications, the Schema Registry and the Message Catalog
+might be managed separately, with Message Catalog entries always referring to
+data type definitions held in the Schema Registry. In simple applications, all
+data type definitions for payloads MAY be held inline inside the Message
+Catalog.
+
+The message definition entries in the catalog contain some fixed metadata
+information. That fixed information MUST uniquely identify the type of message
+and depends on the particular message information model. For example, for CNCF
+CloudEvents message definitions, the "type" attribute is REQUIRED as fixed
+information for each entry.
+
+- Message definition group: A named collection of message definitions. Each
```

Comment:
my first thought on groups is you might want to have an event in many groups, and an easy way do to that would be to make group hierarchy? like trees or some such thing.